src/serverExplorer.ts

```diff
@@ -345,43 +338,68 @@ export class ServersViewTreeDataProvider implements TreeDataProvider< Protocol.S
             const runState: string = this.runStateEnum.get(state.state);
             const pubState: string = this.publishStateEnum.get(state.publishState);
             const depStr = `${id1} (${runState}) (${pubState})`;
-            const treeItem: TreeItem = new TreeItem(`${depStr}`, TreeItemCollapsibleState.Expanded);
-            treeItem.iconPath = ServerIcon.get(handle.type);
-            treeItem.contextValue =  runState;
-            return treeItem;
-        } else if ((item as Protocol.DeployableState).reference ) {
+            return { label: `${depStr}`,
+                iconPath: ServerIcon.get(handle.type),
+                contextValue: runState,
+                collapsibleState: TreeItemCollapsibleState.Expanded
+            };
+        } else if (this.isDeployableElement(item)) {
             const state: Protocol.DeployableState = item as Protocol.DeployableState;
             const id1: string = state.reference.label;
             const runState: string = this.runStateEnum.get(state.state);
             const pubState: string = this.publishStateEnum.get(state.publishState);
             const depStr = `${id1} (${runState}) (${pubState})`;
-            const treeItem: TreeItem = new TreeItem(`${depStr}`);
-            treeItem.iconPath = Uri.file(path.join(__dirname, '../../images/server-light.png'));
-            treeItem.contextValue =  pubState;
-            return treeItem;
+            return { label: `${depStr}`,
+                iconPath: Uri.file(path.join(__dirname, '../../images/server-light.png')),
+                contextValue: pubState,
+                collapsibleState: TreeItemCollapsibleState.None
+            };
+        } else {
+            return undefined;
         }
     }
 
     public getChildren(element?:  Protocol.ServerState | Protocol.DeployableState):  Protocol.ServerState[] | Protocol.DeployableState[] {
         if (element === undefined) {
+            // no parent, root node -> return servers
             return Array.from(this.serverStatus.values());
-        }
-        if ((element as Protocol.ServerState).deployableStates ) {
+        } else if (this.isServerElement(element)) {
+            // server parent -> return deployables
             return (element as Protocol.ServerState).deployableStates;
+        } else {
+            return [];
         }
     }
 
-    public getParent(item?:  Protocol.ServerState | Protocol.DeployableState): Protocol.ServerState | Protocol.DeployableState {
-        if (item === undefined || (item as Protocol.ServerState).deployableStates) {
+    public getParent(element?:  Protocol.ServerState | Protocol.DeployableState): Protocol.ServerState | Protocol.DeployableState {
+        if (element === undefined || this.isServerElement(element)) {
+            return undefined;
+        } else if (this.isDeployableElement(element)) {
+            return this.getServerState(element as Protocol.DeployableState);
+        } else {
             return undefined;
```

Comment:
Maybe we can rewrite it 
if (this.isDeployableElement(element)) {
            return this.getServerState(element as Protocol.DeployableState);
        } else {
            return undefined;